[NVIDIA TF] Optimize embedding_lookup_sparse using new grad op [PART 2/3]

benbarsdell opened the pull request:
Sequel to https://github.com/tensorflow/tensorflow/pull/60383

(This is a rebased version of https://github.com/benbarsdell/tensorflow/pull/1)

cc @cantonios @nluehr @pjannaty

benbarsdell committed:
Use explicit Tweight in segment reduce GPU kernels

- No functional change.

benbarsdell committed:
Use explicit Tindices in segment reduce GPU kernel

- No functional change.

benbarsdell committed:
Add SparseSegmentReduceGradV2 GPU implem

benbarsdell committed:
Rename template params in segment reduce kernels

- Tindex -> Toffsets
- Tweight -> Tweights
- Tindices_internal -> TindicesCompact
- No functional change.

pjannaty commented:
Hi @cantonios, would you be able to help review this PR please?

cantonios commented on tensorflow/core/kernels/segment_reduction_ops_gpu.cu.h:
```diff
@@ -1029,6 +1053,361 @@ struct SparseSegmentGradFunctor<GPUDevice, T, Index, SegmentId> {
+
+    se::Stream* stream = context->op_device_context()->stream();
+    OP_REQUIRES_ASYNC(context, stream,
+                      errors::Internal("No GPU stream available."), done);
```
These errors have since been deprecated, and now we're supposed to use `absl::` versions, i.e. `absl::InternalError(...)`.  For concatenating error messages, this needs to be combined with `absl::StrCat(...)`.  Here and throughout.

benbarsdell committed:
Update deprecated APIs in segment reduce kernels

- Changes errors::XX to absl::XX.
- Replaces GPU_1D_KERNEL_LOOP with GpuGridRangeX.

benbarsdell commented on tensorflow/core/kernels/segment_reduction_ops_gpu.cu.h:
```diff
@@ -29,8 +30,15 @@ limitations under the License.
+#include "tensorflow/core/util/gpu_solvers.h"  // For ScratchSpace
 #include "tensorflow/core/util/permutation_input_iterator.h"
 
+#if GOOGLE_CUDA
```
It's already used like this in [the same file](https://github.com/benbarsdell/tensorflow/blob/12e522e161c39cb4c89d1be152f392b80db469f4/tensorflow/core/kernels/segment_reduction_ops_gpu.cu.h#L19C39-L19C39) as well as hundreds of other places in the codebase.

cantonios replied:
Yes, and we get compile warning for all of those.  So for new instances please first check that the macro exists.

benbarsdell replied:
Done (though I gotta say I don't really understand the motivation).

cantonios replied:
Motivation to reduce compiler warnings?  Because it floods our build logs.

benbarsdell replied:
The thing I don't understand is: if this were a problem, wouldn't all the other instances have been fixed already? It just feels odd to me to be inconsistent, especially within the same file (though I admit that this isn't the first new code vs. old code issue I've run into in TF).

benbarsdell commented on tensorflow/core/kernels/segment_reduction_ops_gpu.cu.h:
```diff
@@ -29,8 +30,15 @@ limitations under the License.
+#if GOOGLE_CUDA
+#include "tensorflow/compiler/xla/stream_executor/cuda/cuda_activation.h"
+#elif TENSORFLOW_USE_ROCM
+#include "tensorflow/core/platform/rocm.h"
```
Not by me, but the [CI passed](http://ml-ci.amd.com:21096/blue/organizations/jenkins/tensorflow%2Fgithub-prs-upstream-master%2FAMD-ROCm-Community-CI-Build/detail/PR-61294/1/pipeline). (This specific series of lines appears in many files such as [here](https://github.com/tensorflow/tensorflow/blob/96375ef/tensorflow/core/common_runtime/gpu/gpu_device.cc#L76-L78) and [here](https://github.com/tensorflow/tensorflow/blob/96375efe2eea4f6baabc65e9f99f1f3b0bb955d2/tensorflow/core/kernels/check_numerics_op.cc#L37-L39)).

benbarsdell commented on tensorflow/core/kernels/segment_reduction_ops_gpu.cu.h:
```diff
@@ -313,34 +321,36 @@ __device__ T ReduceBlockAlongCols(ReduceOp reduce_op, const T& value,
+    Toffsets nouter, Toffsets ninner_vec, Tsegmentids nsegments,
+    ReduceOp reduce_op, Tinit initial_value, Tinit empty_segment_value,
+    bool is_mean, bool is_sqrtn,
+    const Tvec* __restrict__ input_vec,  // [nouter or any, ninner_vec]
```
This is how clang-format put it (otherwise the line becomes too long).

benbarsdell commented on tensorflow/core/kernels/segment_reduction_ops_gpu.cu.h:
```diff
@@ -1029,6 +1053,361 @@ struct SparseSegmentGradFunctor<GPUDevice, T, Index, SegmentId> {
+    const TindicesCompact* __restrict__ sorted_indices,  // [nouter]
+    const Toffsets* __restrict__ sorted_indices_ids,     // [nouter]
+    Tindices* __restrict__ sorted_unique_indices) {      // [num_unique]
+  GPU_1D_KERNEL_LOOP(i, nouter) {
```
Fixed.

benbarsdell committed:
Avoid compiler warning in segment reduce file

benbarsdell committed:
Remove unused variable in segment reduce kernels

benbarsdell commented on tensorflow/core/kernels/segment_reduction_ops_gpu.cu.h:
```diff
@@ -1029,6 +1053,361 @@ struct SparseSegmentGradFunctor<GPUDevice, T, Index, SegmentId> {
+            const Tsegmentids_internal* segment_ids,
+            const TensorShape& dense_output_shape,
+            typename AsyncOpKernel::DoneCallback done) {
+    const GPUDevice& device = context->eigen_gpu_device();
```
Sorry for the delay, fixed.
